### DISCUSSION TITLE
Add implicit viscous joint damping

### DISCUSSION THREAD

**Discussion by hsu** [bulletphysics/bullet3#3479] [General]
Support viscous joint damping using a constraint on the joint's DOF with zero ERP and finite CFM.  References:
- Conversion functions [KpKdToCFMERP or CFMERPToKpKd](https://bitbucket.org/osrf/gazebo/src/783d80d6567cebb1d350296145a87f311acfb38f/gazebo/physics/ode/ODEJoint.cc?at=default#cl-962)
- [joint damping test](https://bitbucket.org/osrf/gazebo/src/783d80d6567cebb1d350296145a87f311acfb38f/test/integration/physics.cc?at=default#cl-686) using [this world](https://bitbucket.org/osrf/gazebo/src/783d80d6567cebb1d350296145a87f311acfb38f/test/worlds/damp_test.world?at=default)

**Comment by scpeters** [comment]
One tricky aspect for us was switching the joint damping constraint on and off at joint limits (see [ODEJoint::ApplyImplicitStiffnessDamping](https://bitbucket.org/osrf/gazebo/src/783d80d6567cebb1d350296145a87f311acfb38f/gazebo/physics/ode/ODEJoint.cc#cl-842)).

**Comment by erwincoumans** [comment]
I won't have time to implement the feature for Bullet 2.83. I wonder how your damping (and spring) compares with the btGeneric6DofSpring2Constraint version?
